{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "from os import path,listdir\n",
    "import pandas as pd\n",
    "import csv\n",
    "import subprocess\n",
    "import argparse\n",
    "import configparser\n",
    "from collections import OrderedDict\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import enum\n",
    "\n",
    "#~~~~~~~~HELPERS FOR BUILDING PYGT CONFIG FILE\n",
    "#~~~Gencode annotation\n",
    "str_track_gtf=\"\"\"[x-axis]\n",
    "where=top\n",
    "\n",
    "[test gtf]\n",
    "title={gtf_title}\n",
    "file={gtf_input}\n",
    "file_type=gtf \n",
    "height={gtf_height}\n",
    "#~I think this is a bug in gtf track implementation\n",
    "color=black\n",
    "#~Labels on annotations\n",
    "labels=true\n",
    "fontsize=8\n",
    "prefered_name=transcript_name\n",
    "#~Arrows on TSSs (comment line to remove)\n",
    "style=tssarrow\n",
    "\n",
    "[spacer]\n",
    "height = 0.4\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "str_track_line=\"\"\"\n",
    "[lspacer]\n",
    "file_type=hlines\n",
    "show_data_range=false\n",
    "y_values=.5\n",
    "overlay_previous=share-y\n",
    "height = 0.8\n",
    "\n",
    "#[spacer]\n",
    "#height = 0.1\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "#~~~Group spacer (more than 1 bed file per group)\n",
    "str_track_grouplabel=\"\"\"\n",
    "[bedspacer {group}]\n",
    "file = {emptybed}\n",
    "title = {group}\n",
    "file_type = bed\n",
    "gene_rows = 1\n",
    "height=.8\n",
    "\"\"\"\n",
    "\n",
    "#~~~Group spacer (General bed formatting)\n",
    "str_track_bed=\"\"\"\n",
    "[test {group}-{title}]\n",
    "file = {filename}\n",
    "title = {title}\n",
    "color = {color}\n",
    "labels = false\n",
    "file_type = bed\n",
    "display = collapsed\n",
    "gene_rows = 1\n",
    "arrowhead_included = true\n",
    "style = UCSC\n",
    "height=.8\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "#~~~~~~~~~~~~~~~~~~~~\n",
    "class Mode(enum.Flag):\n",
    "    REGION = enum.auto()\n",
    "    TRANS = enum.auto()\n",
    "    TRANS_NOINTRONS = enum.auto()    \n",
    "    ERROR = enum.auto()\n",
    "    \n",
    "def get_mode(config_obj):\n",
    "    config = config_obj\n",
    "    if 'region' in config['default']:\n",
    "        return Mode.REGION\n",
    "    transcript = 'transcript' in config['default']\n",
    "    no_introns = config.getboolean('default','no_introns',fallback=False)    \n",
    "    if transcript and no_introns:\n",
    "        return Mode.TRANS_NOINTRONS\n",
    "    if transcript:\n",
    "        return Mode.TRANS\n",
    "    return Mode.ERROR\n",
    "\n",
    "def load_config(config_ini):\n",
    "    if not path.exists(config_ini):\n",
    "        raise FileNotFoundError(\"Config file not found. Exiting.\")\n",
    "    config = configparser.ConfigParser()\n",
    "    config.optionxform = str\n",
    "    config.read(config_ini)\n",
    "    mode = get_mode(config)\n",
    "    return config,mode\n",
    "\n",
    "def load_tracks(config_obj):\n",
    "    config=config_obj\n",
    "    tracks = OrderedDict()\n",
    "    for group in config['tracks']:\n",
    "        P = config['tracks'][group]\n",
    "        if not path.exists(P):\n",
    "            raise FileNotFoundError('ERROR: track group',P,'not found.')\n",
    "        if path.isfile(P):\n",
    "            tracks[group] = [(P,group)]\n",
    "        else:            \n",
    "            files = [path.join(P,x) for x in listdir(P) if x.endswith('.bed') or x.endswith('.bed.gz')]\n",
    "            if len(files)==0:\n",
    "                print('# No bed tracks found in %s. Skipping.',file=sys.stderr)\n",
    "                continue\n",
    "            tracks[group] = []\n",
    "            for p in files:                \n",
    "                p_name = path.basename(p).replace('.bed.gz','').replace('.bed','')                \n",
    "                tracks[group].append((p,p_name))                \n",
    "    return tracks\n",
    "    \n",
    "def print_config(config_obj):\n",
    "    for s in config_obj.sections():\n",
    "        print('[%s]'%s)\n",
    "        for k in config_obj[s]:\n",
    "            print('%s = %s'%(k,config_obj[s][k]))\n",
    "\n",
    "#~~~~~~~~~~~~~~~~~~~~\n",
    "def check(config_ini):\n",
    "    errors = 0\n",
    "    #~~~Check programs\n",
    "    try:\n",
    "        call = 'echo $CONDA_PREFIX;python -V'\n",
    "        call_output = subprocess.check_output(call,shell=True).decode().strip().replace('\\n',': ')\n",
    "        print('# conda_env:',call_output,file=sys.stderr)\n",
    "    except:\n",
    "        print('ERROR: Conda environment not found. Exiting.',file=sys.stderr)\n",
    "        errors+=1\n",
    "    try:\n",
    "        call = 'pyGenomeTracks --version'\n",
    "        call_output = subprocess.check_output(call,shell=True).decode().strip()\n",
    "        print('#',call_output,file=sys.stderr)\n",
    "    except:\n",
    "        print('ERROR: pyGenomeTracks not found. Exiting.',file=sys.stderr)\n",
    "        errors+=1\n",
    "    try:\n",
    "        call = 'bedtools -version'\n",
    "        call_output = subprocess.check_output(call,shell=True).decode().strip()\n",
    "        print('#',call_output,file=sys.stderr)\n",
    "    except:\n",
    "        print('ERROR: bedtools not found. Exiting.',file=sys.stderr)\n",
    "        errors+=1\n",
    "    \n",
    "    #~~~Check configuration\n",
    "    if path.exists(config_ini):\n",
    "        print('# config_file found at',path.abspath(config_ini),file=sys.stderr)\n",
    "    else:\n",
    "        print('ERROR: config_file not found. Exiting.',file=sys.stderr)\n",
    "        return\n",
    "    config, mode = load_config(config_ini)\n",
    "    required_params = ['output_path','input_gtf']\n",
    "    for param in required_params:\n",
    "        if param not in config['default']:\n",
    "            print('ERROR: required parameter %s not found in configuration.'%param,file=sys.stderr)\n",
    "            errors+=1\n",
    "    if mode is Mode.ERROR:\n",
    "        print('ERROR: you need to specify a region or transcript in the configuration.',file=sys.stderr)\n",
    "        errors+=1\n",
    "    else:\n",
    "        print('# MODE %s detected.'%mode,file=sys.stderr)\n",
    "    required_files = ['input_gtf']\n",
    "    for  _file in required_files:\n",
    "        filepath = config['default'][_file]\n",
    "        if not path.exists(filepath):\n",
    "            print('ERROR: required %s (%s) not found.'%(_file,filepath),file=sys.stderr)\n",
    "            errors+=1\n",
    "    \n",
    "    #~~~Check track files\n",
    "    tracks = OrderedDict()\n",
    "    for group in config['tracks']:\n",
    "        P = config['tracks'][group]\n",
    "        if not path.exists(P):\n",
    "            print('ERROR: track group',P,'not found.',file=sys.stderr)\n",
    "            errors+=1\n",
    "            continue\n",
    "        print('# Found %s: %s'%(group,P),file=sys.stderr)\n",
    "        if path.isfile(P):\n",
    "            tracks[group] = [(P,group)]\n",
    "        else:\n",
    "            tracks[group] = []\n",
    "            files = [path.join(P,x) for x in listdir(P) if x.endswith('.bed') or x.endswith('.bed.gz')]\n",
    "            for p in files:                \n",
    "                p_name = path.basename(p).replace('.bed.gz','').replace('.bed','')                \n",
    "                tracks[group].append((p,p_name))\n",
    "                print('# Found %s-%s: %s'%(group,p_name,p),file=sys.stderr)    \n",
    "\n",
    "    #~~~End checkup\n",
    "    print('# Checkup finalized. %d error(s) detected.'%errors,file=sys.stderr)\n",
    "    if errors>0:\n",
    "        print('# Please solve all the configuration errors before calling \"prepare\" and \"draw\".',file=sys.stderr)    \n",
    "    \n",
    "#~~~~~~~~~~~~~~~~~~~~\n",
    "def execute_calls(calls):\n",
    "    for call in calls:        \n",
    "        print(call,file=sys.stderr)\n",
    "        result = subprocess.run(call,shell=True,stdout=subprocess.PIPE)\n",
    "        msg = result.stdout.decode().strip()\n",
    "        if msg!='':\n",
    "            print(msg,file=sys.stderr)\n",
    "            \n",
    "#~~~~~~~~~~~~~~~~~~~~\n",
    "def create_relative_mapping_index(prep_path):\n",
    "    #~~~Inputs\n",
    "    prep_gtf = path.join(prep_path,'input.gtf')\n",
    "    query_bed = path.join(prep_path,'query.bed')\n",
    "    #~~~Outputs\n",
    "    index_bed = path.join(prep_path,'index.bed')\n",
    "    vert_bed = path.join(prep_path,'vertical.bed')\n",
    "    #~~~Prepare index (chr,start,end,intron_length_to_subtract)\n",
    "    df = pd.read_csv(query_bed,sep='\\t',header=None)    \n",
    "    X = [df.iloc[0][1]]+list(df[2].values)[:-1]\n",
    "    df[3] = X\n",
    "    df['length'] = df[2]-df[1]\n",
    "    df['intron'] = df[1]-df[3]\n",
    "    df['cumsum'] = df['intron'].cumsum()\n",
    "    df[[0,1,2,'cumsum']].to_csv(index_bed,sep='\\t',header=None,index=None)\n",
    "    t_length = df['length'].sum()\n",
    "    #~~~Prepare gtf\n",
    "    dgt = pd.read_csv(prep_gtf,sep='\\t',header=None)\n",
    "    offset_gtf = dgt[3].min()\n",
    "    dfexons = dgt[dgt[2]=='exon'].sort_values(by=3)\n",
    "    dftrans = dgt[dgt[2]=='transcript'].copy()\n",
    "    dfexons[3] = (dfexons[3]-df['cumsum'].values)-offset_gtf+1\n",
    "    dfexons[4] = (dfexons[4]-df['cumsum'].values)-offset_gtf+1\n",
    "    dfexons['bed'] = dfexons[3]-1\n",
    "    dfexons[[0,'bed',3]].to_csv(vert_bed,sep='\\t',index=None,header=None)\n",
    "    dfexons.drop(columns=['bed'],inplace=True)\n",
    "    with open(vert_bed,'a') as f:\n",
    "        f.write('\\t'.join([dfexons.iloc[0][0],str(t_length-1),str(t_length)])+'\\n')\n",
    "    dftrans[3] = 1\n",
    "    dftrans[4] = t_length\n",
    "    dgt = pd.concat((dftrans,dfexons)).sort_index()\n",
    "    dgt.to_csv(prep_gtf,sep='\\t',index=None,header=None,quoting=csv.QUOTE_NONE)\n",
    "    offset_bed = offset_gtf-1\n",
    "    return offset_bed\n",
    "\n",
    "def transcript_info(input_gtf):\n",
    "    df = pd.read_csv(input_gtf,sep='\\t',header=None)\n",
    "    T = list(df[df[2]=='transcript'].to_dict(orient='index').values())[0]\n",
    "    R = {'chr':T[0],'start':T[3]-1,'end':T[4],'strand':T[6],\n",
    "         'length':T[4]-T[3]+1,'midpoint':(T[4]+T[3]-1)/2}\n",
    "    return R\n",
    "    \n",
    "#~~~~~~~~~~~~~~~~~~~~\n",
    "def prepare(config_ini):\n",
    "    config, mode = load_config(config_ini)\n",
    "    output_path = path.abspath(config['default']['output_path'])\n",
    "    input_gtf = config['default']['input_gtf']\n",
    "    prep_path = path.join(output_path,'prep')\n",
    "    prep_gtf = path.join(prep_path,'input.gtf')\n",
    "    query_bed = path.join(prep_path,'query.bed')\n",
    "    folders = [output_path,prep_path]\n",
    "    reverse_mode = False\n",
    "    for folder in folders:\n",
    "        if not path.exists(folder):\n",
    "            os.makedirs(folder)\n",
    "            print('# Created dir %s'%folder,file=sys.stderr)\n",
    "    calls = []\n",
    "    call_intersect = 'intersectBed -sorted -u -a {input_bed} -b {query_bed} > {output_bed}'\n",
    "    if mode is Mode.REGION:        \n",
    "        call_gtf = 'intersectBed -sorted -u -a {input_gtf} -b {query_bed} | sort -k1,1 -k4,4n > {prep_gtf}'            \n",
    "        region = config['default']['region']\n",
    "        print('# REGION MODE: %s'%region,file=sys.stderr)\n",
    "        _chr = region.split(':')[0]\n",
    "        start = str(int(region.split(':')[1].split('-')[0])-1)\n",
    "        end = region.split(':')[1].split('-')[1]        \n",
    "        with open(query_bed,'w') as f:            \n",
    "            f.write('\\t'.join([_chr,start,end])+'\\n')\n",
    "            print('# Wrote',query_bed,file=sys.stderr)    \n",
    "        call = call_gtf.format(query_bed=query_bed,input_gtf=input_gtf,prep_gtf=prep_gtf)\n",
    "        calls.append(call)\n",
    "    elif mode & (Mode.TRANS|Mode.TRANS_NOINTRONS):        \n",
    "        transcript = config['default']['transcript']\n",
    "        call_gtf_search = '{compressed}grep {transcript} {gtf} | sort -k1,1 -k4,4n > {prep_gtf}'\n",
    "        call1 = call_gtf_search.format(transcript=transcript,gtf=input_gtf,prep_gtf=prep_gtf,\n",
    "                                     compressed='z' if input_gtf.endswith('.gz') else '')        \n",
    "        call_bed = 'grep -P \"\\\\t{feature}\\\\t\" {prep_gtf} | awk \\'BEGIN{{OFS=FS=\"\\\\t\"}}{{print $1,$4-1,$5}}\\' | sort -k1,1 -k2,2n > {query_bed}'\n",
    "        call2 = call_bed.format(feature='exon' if mode is Mode.TRANS_NOINTRONS else 'transcript',\n",
    "                               prep_gtf=prep_gtf,query_bed=query_bed)\n",
    "        execute_calls([call1,call2])        \n",
    "        if mode is Mode.TRANS_NOINTRONS:\n",
    "            offset_bed = create_relative_mapping_index(prep_path)\n",
    "            call_intersect = 'intersectBed -sorted -wb -a {input_bed} -b {query_bed} '\n",
    "            call_intersect+= ' | awk \\'BEGIN{{OFS=FS=\"\\\\t\"}}{{printf $1FS$2-$NF-%dFS$3-$NF-%d;for(i=4;i<=NF-4;i++) printf FS$i;print \"\"}}\\' > {output_bed}'%(offset_bed,offset_bed)\n",
    "            index_bed = path.join(prep_path,'index.bed')\n",
    "            query_bed = index_bed\n",
    "        #~~~Prepare for reversing\n",
    "        TI = transcript_info(prep_gtf)\n",
    "        force_forward = config.getboolean('default','force_forward',fallback=False) \n",
    "        reverse_mode = force_forward and TI['strand']=='-'\n",
    "        reverse_script = path.join(os.getcwd(),'reverseBed.sh')\n",
    "        call_reverse = reverse_script+' -i {output_bed} -c %.1f > {output_bed}ff; mv {output_bed}ff {output_bed}'%TI['midpoint']\n",
    "    else:\n",
    "        raise AttributeError(\"Invalid mode, please check configuration file.\")\n",
    "    TRACKS = load_tracks(config)\n",
    "    for group,tracks in TRACKS.items():\n",
    "        track_path = path.join(prep_path,group)\n",
    "        if not path.exists(track_path):\n",
    "            os.makedirs(track_path)\n",
    "            print('# Created dir %s'%track_path,file=sys.stderr)\n",
    "        for track in tracks:\n",
    "            output_bed = path.join(track_path,track[1]+'.bed')            \n",
    "            call = call_intersect.format(query_bed=query_bed,input_bed=track[0],output_bed=output_bed)\n",
    "            calls.append(call)\n",
    "            if reverse_mode:\n",
    "                calls.append(call_reverse.format(output_bed=output_bed))\n",
    "    if reverse_mode:\n",
    "        calls.append(reverse_script+' -i {gtf} > {gtf}ff; mv {gtf}ff {gtf}'.format(gtf=prep_gtf))\n",
    "    execute_calls(calls)\n",
    "    print('# Done.',file=sys.stderr)\n",
    "    \n",
    "#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "call_PGT = 'pyGenomeTracks --tracks {output_ini} --region {chr}:{view_start}-{view_end} --fontSize 12 --trackLabelFraction 0.2 --trackLabelHAlign left --width {width} --dpi {dpi} -o {output_img}'\n",
    "\n",
    "def draw(config_ini):\n",
    "    config, mode = load_config(config_ini)\n",
    "    #~~~\n",
    "    relative_mode = False\n",
    "    if 'transcript' in config['default'] and config.getboolean('default','relative',fallback=False):\n",
    "        relative_mode = True\n",
    "    #~~~\n",
    "    output_path = path.abspath(config['default']['output_path'])\n",
    "    prep_path = path.join(output_path,'prep')\n",
    "    prep_gtf = path.join(prep_path,'input.gtf')\n",
    "    output_ini = path.join(output_path,'config.ini')    \n",
    "    output_img = path.join(output_path, config['plot'].get('output_img','my_tracks.pdf'))\n",
    "    empty_bed = path.join(prep_path,'empty.bed')\n",
    "    with open(empty_bed,'w') as f:\n",
    "        pass\n",
    "    \n",
    "    #~~~Graphics parameters\n",
    "    gtf_title = config['plot'].get('gtf_title','Annotation')\n",
    "    padding = int(config['plot'].get('padding','100'))    \n",
    "    width = config['plot'].get('width','40')\n",
    "    dpi = config['plot'].get('dpi','100')\n",
    "    gtf_height = config['plot'].get('gtf_height','10')\n",
    "    \n",
    "    #~~~Get plotting coordinates\n",
    "    if mode is Mode.REGION:    \n",
    "        region = config['default']['region']\n",
    "        print('# REGION MODE: %s'%region,file=sys.stderr)\n",
    "        _chr = region.split(':')[0]\n",
    "        start = int(region.split(':')[1].split('-')[0])-1\n",
    "        end = int(region.split(':')[1].split('-')[1])\n",
    "        view_start = start-padding\n",
    "        view_end = end+padding\n",
    "    elif mode & (Mode.TRANS|Mode.TRANS_NOINTRONS):     \n",
    "        transcript = config['default']['transcript']                        \n",
    "        df = pd.read_csv(prep_gtf,sep='\\t',header=None)\n",
    "        _chr = df.iloc[0][0]\n",
    "        start = df[3].min()-1\n",
    "        end = df[4].max()\n",
    "        view_start = max(start-padding,0)\n",
    "        view_end = end+padding\n",
    "    else:\n",
    "        raise AttributeError(\"Invalid mode, please check configuration file.\")\n",
    "    #~~~ Generate HTML color list\n",
    "    cmap = plt.get_cmap('tab10')\n",
    "    colors = [matplotlib.colors.rgb2hex(cmap(i)[:3]) for i in range(cmap.N)]\n",
    "    \n",
    "    #~~~Create ini file (GTF)\n",
    "    with open(output_ini,'w') as f:\n",
    "        f.write(str_track_gtf.format(gtf_input=prep_gtf,gtf_title=gtf_title,gtf_height=gtf_height))\n",
    "        if mode is Mode.TRANS_NOINTRONS:\n",
    "            f.write(\"[vlines]\\nfile = {file}\\ntype = vlines\\n\".format(file=path.join(prep_path,'vertical.bed')))\n",
    "    \n",
    "    #~~~Add tracks to ini file\n",
    "    TRACKS = load_tracks(config)\n",
    "    track_count = 0\n",
    "    with open(output_ini,'a') as f:\n",
    "        for idx,(group,tracks) in enumerate(TRACKS.items()):\n",
    "            color = colors[idx%10]\n",
    "            track_path = path.join(prep_path,group)\n",
    "            if len(tracks) == 1: #Only one track in the group\n",
    "                filename = path.join(track_path,tracks[0][1]+'.bed')\n",
    "                if os.stat(filename).st_size == 0:\n",
    "                    continue\n",
    "                f.write(str_track_bed.format(group=group,color=color,\n",
    "                                        filename=filename,\n",
    "                                        title=group))\n",
    "                f.write(str_track_line)\n",
    "                track_count+=1\n",
    "            else:\n",
    "                f.write(str_track_grouplabel.format(group=group,emptybed=empty_bed))\n",
    "                for track in tracks:                    \n",
    "                    filename = path.join(track_path,track[1]+'.bed')                    \n",
    "                    if os.stat(filename).st_size == 0:\n",
    "                        continue\n",
    "                    f.write(str_track_bed.format(group=group,color=color,\n",
    "                                                 filename=filename,\n",
    "                                                 title=track[1]))\n",
    "                    f.write(str_track_line)\n",
    "                    track_count+=1                    \n",
    "    call = call_PGT.format(output_ini=output_ini,chr=_chr,view_start=view_start,view_end=view_end,\n",
    "                         output_img=output_img,width=width,dpi=dpi)    \n",
    "    result = subprocess.run(call,shell=True,stderr=subprocess.PIPE)\n",
    "    msg = result.stderr.decode().strip()\n",
    "    if msg!='':\n",
    "        print(msg,file=sys.stderr)\n",
    "    print('# Wrote',output_img,file=sys.stderr)\n",
    "    print('#######################################################################',file=sys.stderr)\n",
    "    print('### If you want to customize the image modify %s and then rerun\\n%s'%(path.abspath(output_ini),call),file=sys.stderr)\n",
    "    print('#######################################################################',file=sys.stderr)\n",
    "\n",
    "    \n",
    "    \n",
    "def main():\n",
    "    parser = argparse.ArgumentParser(description='ezTracks. Plot a single GTF annotation followed by grouped bed files.\\n\\nHugo Guillen, 2020.')\n",
    "    \n",
    "    parser.add_argument('action',choices=['check','prepare','draw'],help=\"\"\"check: check if config file, environment, and input files are correct.\n",
    "prepare: preprocess input files into output defined in config file.\n",
    "draw: generate output image into output defined in config file.\"\"\")\n",
    "    parser.add_argument('configfile',help=\"path to configuration file. For possible options, please refer to the example located at test_data/config.ini\")\n",
    "    if len(sys.argv) < 2:\n",
    "        parser.print_help()\n",
    "        sys.exit(1)    \n",
    "    args = parser.parse_args()    \n",
    "    if args.action == 'check': \n",
    "        check(args.configfile)\n",
    "    elif args.action == 'prepare':\n",
    "        prepare(args.configfile)\n",
    "    elif args.action == 'draw':\n",
    "        draw(args.configfile)\n",
    "    else:\n",
    "        parser.print_help()\n",
    "        \n",
    "# if __name__ == \"__main__\":\n",
    "#     main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "# REGION MODE: chrX:73816525-73859248\n",
      "INFO:pygenometracks.tracksClass:initialize 1. [x-axis]\n",
      "INFO:pygenometracks.tracksClass:initialize 2. [test gtf]\n",
      "100%|██████████| 14/14 [00:00<00:00, 2742.27it/s]\n",
      "INFO:pygenometracks.tracksClass:initialize 3. [spacer]\n",
      "INFO:pygenometracks.tracksClass:initialize 4. [test CRS-CRS]\n",
      "100%|██████████| 29/29 [00:00<00:00, 33891.00it/s]\n",
      "INFO:pygenometracks.tracksClass:initialize 5. [lspacer]\n",
      "INFO:pygenometracks.tracksClass:initialize 6. [bedspacer phastCons]\n",
      "0it [00:00, ?it/s]\n",
      "WARNING:pygenometracks.tracks.GenomeTrack:No valid intervals were found in file /home/hugo/software/ezTracks/test_output/test_region/prep/empty.bed\n",
      "INFO:pygenometracks.tracksClass:initialize 7. [test phastCons-20way]\n",
      "100%|██████████| 80/80 [00:00<00:00, 34267.19it/s]\n",
      "INFO:pygenometracks.tracksClass:initialize 8. [lspacer]\n",
      "INFO:pygenometracks.tracksClass:initialize 9. [test phastCons-7way]\n",
      "100%|██████████| 26/26 [00:00<00:00, 35326.18it/s]\n",
      "INFO:pygenometracks.tracksClass:initialize 10. [lspacer]\n",
      "INFO:pygenometracks.tracksClass:initialize 11. [bedspacer repeats]\n",
      "0it [00:00, ?it/s]\n",
      "WARNING:pygenometracks.tracks.GenomeTrack:No valid intervals were found in file /home/hugo/software/ezTracks/test_output/test_region/prep/empty.bed\n",
      "INFO:pygenometracks.tracksClass:initialize 12. [test repeats-Alu]\n",
      "100%|██████████| 15/15 [00:00<00:00, 32082.90it/s]\n",
      "INFO:pygenometracks.tracksClass:initialize 13. [lspacer]\n",
      "INFO:pygenometracks.tracksClass:initialize 14. [test repeats-Simple]\n",
      "100%|██████████| 8/8 [00:00<00:00, 26715.31it/s]\n",
      "INFO:pygenometracks.tracksClass:initialize 15. [lspacer]\n",
      "INFO:pygenometracks.tracksClass:time initializing track(s):\n",
      "INFO:pygenometracks.tracksClass:0.05416369438171387\n",
      "DEBUG:pygenometracks.tracksClass:Figure size in cm is 40.0 x 19.015957446808514. Dpi is set to 100\n",
      "\n",
      "INFO:pygenometracks.tracksClass:plotting 1. [x-axis]\n",
      "INFO:pygenometracks.tracksClass:plotting 2. [test gtf]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:ylim 26.379999999999995,-0.08\n",
      "INFO:pygenometracks.tracksClass:plotting 3. [spacer]\n",
      "INFO:pygenometracks.tracksClass:plotting 4. [test CRS-CRS]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:ylim 2.3,-0.08\n",
      "INFO:pygenometracks.tracksClass:plotting 5. [lspacer]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_values: [0.5]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_value: 0.5\n",
      "INFO:pygenometracks.tracksClass:plotting 6. [bedspacer phastCons]\n",
      "WARNING:pygenometracks.tracks.GenomeTrack:*Warning*\n",
      "Neither chrX nor X existss as a chromosome name inside the bed file. This will generate an empty track!!\n",
      "\n",
      "INFO:pygenometracks.tracksClass:plotting 7. [test phastCons-20way]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:ylim 2.3,-0.08\n",
      "INFO:pygenometracks.tracksClass:plotting 8. [lspacer]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_values: [0.5]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_value: 0.5\n",
      "INFO:pygenometracks.tracksClass:plotting 9. [test phastCons-7way]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:ylim 2.3,-0.08\n",
      "INFO:pygenometracks.tracksClass:plotting 10. [lspacer]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_values: [0.5]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_value: 0.5\n",
      "INFO:pygenometracks.tracksClass:plotting 11. [bedspacer repeats]\n",
      "WARNING:pygenometracks.tracks.GenomeTrack:*Warning*\n",
      "Neither chrX nor X existss as a chromosome name inside the bed file. This will generate an empty track!!\n",
      "\n",
      "INFO:pygenometracks.tracksClass:plotting 12. [test repeats-Alu]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:ylim 2.3,-0.08\n",
      "INFO:pygenometracks.tracksClass:plotting 13. [lspacer]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_values: [0.5]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_value: 0.5\n",
      "INFO:pygenometracks.tracksClass:plotting 14. [test repeats-Simple]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:ylim 2.3,-0.08\n",
      "INFO:pygenometracks.tracksClass:plotting 15. [lspacer]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_values: [0.5]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_value: 0.5\n",
      "# Wrote /home/hugo/software/ezTracks/test_output/test_region/my_tracks.png\n",
      "#######################################################################\n",
      "### If you want to customize the image modify /home/hugo/software/ezTracks/test_output/test_region/config.ini and then rerun\n",
      "pyGenomeTracks --tracks /home/hugo/software/ezTracks/test_output/test_region/config.ini --region chrX:73816424-73859348 --fontSize 12 --trackLabelFraction 0.2 --trackLabelHAlign left --width 40 --dpi 100 -o /home/hugo/software/ezTracks/test_output/test_region/my_tracks.png\n",
      "#######################################################################\n",
      "INFO:pygenometracks.tracksClass:initialize 1. [x-axis]\n",
      "INFO:pygenometracks.tracksClass:initialize 2. [test gtf]\n",
      "100%|██████████| 1/1 [00:00<00:00, 1368.01it/s]\n",
      "INFO:pygenometracks.tracksClass:initialize 3. [spacer]\n",
      "INFO:pygenometracks.tracksClass:initialize 4. [test CRS-CRS]\n",
      "100%|██████████| 29/29 [00:00<00:00, 35944.09it/s]\n",
      "INFO:pygenometracks.tracksClass:initialize 5. [lspacer]\n",
      "INFO:pygenometracks.tracksClass:initialize 6. [bedspacer phastCons]\n",
      "0it [00:00, ?it/s]\n",
      "WARNING:pygenometracks.tracks.GenomeTrack:No valid intervals were found in file /home/hugo/software/ezTracks/test_output/test_trans/prep/empty.bed\n",
      "INFO:pygenometracks.tracksClass:initialize 7. [test phastCons-20way]\n",
      "100%|██████████| 75/75 [00:00<00:00, 36514.54it/s]\n",
      "INFO:pygenometracks.tracksClass:initialize 8. [lspacer]\n",
      "INFO:pygenometracks.tracksClass:initialize 9. [test phastCons-7way]\n",
      "100%|██████████| 24/24 [00:00<00:00, 31011.49it/s]\n",
      "INFO:pygenometracks.tracksClass:initialize 10. [lspacer]\n",
      "INFO:pygenometracks.tracksClass:initialize 11. [bedspacer repeats]\n",
      "0it [00:00, ?it/s]\n",
      "WARNING:pygenometracks.tracks.GenomeTrack:No valid intervals were found in file /home/hugo/software/ezTracks/test_output/test_trans/prep/empty.bed\n",
      "INFO:pygenometracks.tracksClass:initialize 12. [test repeats-Alu]\n",
      "100%|██████████| 5/5 [00:00<00:00, 20805.08it/s]\n",
      "INFO:pygenometracks.tracksClass:initialize 13. [lspacer]\n",
      "INFO:pygenometracks.tracksClass:initialize 14. [test repeats-Simple]\n",
      "100%|██████████| 7/7 [00:00<00:00, 23450.58it/s]\n",
      "INFO:pygenometracks.tracksClass:initialize 15. [lspacer]\n",
      "INFO:pygenometracks.tracksClass:time initializing track(s):\n",
      "INFO:pygenometracks.tracksClass:0.027378559112548828\n",
      "DEBUG:pygenometracks.tracksClass:Figure size in cm is 25.0 x 9.97340425531915. Dpi is set to 100\n",
      "\n",
      "INFO:pygenometracks.tracksClass:plotting 1. [x-axis]\n",
      "INFO:pygenometracks.tracksClass:plotting 2. [test gtf]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:ylim 1.08,-0.08\n",
      "INFO:pygenometracks.tracksClass:plotting 3. [spacer]\n",
      "INFO:pygenometracks.tracksClass:plotting 4. [test CRS-CRS]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:ylim 2.3,-0.08\n",
      "INFO:pygenometracks.tracksClass:plotting 5. [lspacer]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_values: [0.5]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_value: 0.5\n",
      "INFO:pygenometracks.tracksClass:plotting 6. [bedspacer phastCons]\n",
      "WARNING:pygenometracks.tracks.GenomeTrack:*Warning*\n",
      "Neither chrX nor X existss as a chromosome name inside the bed file. This will generate an empty track!!\n",
      "\n",
      "INFO:pygenometracks.tracksClass:plotting 7. [test phastCons-20way]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:ylim 2.3,-0.08\n",
      "INFO:pygenometracks.tracksClass:plotting 8. [lspacer]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_values: [0.5]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_value: 0.5\n",
      "INFO:pygenometracks.tracksClass:plotting 9. [test phastCons-7way]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:ylim 2.3,-0.08\n",
      "INFO:pygenometracks.tracksClass:plotting 10. [lspacer]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_values: [0.5]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_value: 0.5\n",
      "INFO:pygenometracks.tracksClass:plotting 11. [bedspacer repeats]\n",
      "WARNING:pygenometracks.tracks.GenomeTrack:*Warning*\n",
      "Neither chrX nor X existss as a chromosome name inside the bed file. This will generate an empty track!!\n",
      "\n",
      "INFO:pygenometracks.tracksClass:plotting 12. [test repeats-Alu]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:ylim 2.3,-0.08\n",
      "INFO:pygenometracks.tracksClass:plotting 13. [lspacer]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_values: [0.5]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_value: 0.5\n",
      "INFO:pygenometracks.tracksClass:plotting 14. [test repeats-Simple]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:ylim 2.3,-0.08\n",
      "INFO:pygenometracks.tracksClass:plotting 15. [lspacer]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_values: [0.5]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_value: 0.5\n",
      "# Wrote /home/hugo/software/ezTracks/test_output/test_trans/my_tracks.png\n",
      "#######################################################################\n",
      "### If you want to customize the image modify /home/hugo/software/ezTracks/test_output/test_trans/config.ini and then rerun\n",
      "pyGenomeTracks --tracks /home/hugo/software/ezTracks/test_output/test_trans/config.ini --region chrX:73820555-73852823 --fontSize 12 --trackLabelFraction 0.2 --trackLabelHAlign left --width 25 --dpi 100 -o /home/hugo/software/ezTracks/test_output/test_trans/my_tracks.png\n",
      "#######################################################################\n",
      "INFO:pygenometracks.tracksClass:initialize 1. [x-axis]\n",
      "INFO:pygenometracks.tracksClass:initialize 2. [test gtf]\n",
      "100%|██████████| 1/1 [00:00<00:00, 1220.34it/s]\n",
      "INFO:pygenometracks.tracksClass:initialize 3. [spacer]\n",
      "INFO:pygenometracks.tracksClass:initialize 4. [test CRS-CRS]\n",
      "100%|██████████| 29/29 [00:00<00:00, 36648.03it/s]\n",
      "INFO:pygenometracks.tracksClass:initialize 5. [lspacer]\n",
      "INFO:pygenometracks.tracksClass:initialize 6. [bedspacer phastCons]\n",
      "0it [00:00, ?it/s]\n",
      "WARNING:pygenometracks.tracks.GenomeTrack:No valid intervals were found in file /home/hugo/software/ezTracks/test_output/test_trans_ff/prep/empty.bed\n",
      "INFO:pygenometracks.tracksClass:initialize 7. [test phastCons-20way]\n",
      "100%|██████████| 75/75 [00:00<00:00, 37768.38it/s]\n",
      "INFO:pygenometracks.tracksClass:initialize 8. [lspacer]\n",
      "INFO:pygenometracks.tracksClass:initialize 9. [test phastCons-7way]\n",
      "100%|██████████| 24/24 [00:00<00:00, 32587.66it/s]\n",
      "INFO:pygenometracks.tracksClass:initialize 10. [lspacer]\n",
      "INFO:pygenometracks.tracksClass:initialize 11. [bedspacer repeats]\n",
      "0it [00:00, ?it/s]\n",
      "WARNING:pygenometracks.tracks.GenomeTrack:No valid intervals were found in file /home/hugo/software/ezTracks/test_output/test_trans_ff/prep/empty.bed\n",
      "INFO:pygenometracks.tracksClass:initialize 12. [test repeats-Alu]\n",
      "100%|██████████| 5/5 [00:00<00:00, 14799.94it/s]\n",
      "INFO:pygenometracks.tracksClass:initialize 13. [lspacer]\n",
      "INFO:pygenometracks.tracksClass:initialize 14. [test repeats-Simple]\n",
      "100%|██████████| 7/7 [00:00<00:00, 23658.44it/s]\n",
      "INFO:pygenometracks.tracksClass:initialize 15. [lspacer]\n",
      "INFO:pygenometracks.tracksClass:time initializing track(s):\n",
      "INFO:pygenometracks.tracksClass:0.02777886390686035\n",
      "DEBUG:pygenometracks.tracksClass:Figure size in cm is 25.0 x 9.97340425531915. Dpi is set to 100\n",
      "\n",
      "INFO:pygenometracks.tracksClass:plotting 1. [x-axis]\n",
      "INFO:pygenometracks.tracksClass:plotting 2. [test gtf]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:ylim 1.08,-0.08\n",
      "INFO:pygenometracks.tracksClass:plotting 3. [spacer]\n",
      "INFO:pygenometracks.tracksClass:plotting 4. [test CRS-CRS]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:ylim 2.3,-0.08\n",
      "INFO:pygenometracks.tracksClass:plotting 5. [lspacer]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_values: [0.5]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_value: 0.5\n",
      "INFO:pygenometracks.tracksClass:plotting 6. [bedspacer phastCons]\n",
      "WARNING:pygenometracks.tracks.GenomeTrack:*Warning*\n",
      "Neither chrX nor X existss as a chromosome name inside the bed file. This will generate an empty track!!\n",
      "\n",
      "INFO:pygenometracks.tracksClass:plotting 7. [test phastCons-20way]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:ylim 2.3,-0.08\n",
      "INFO:pygenometracks.tracksClass:plotting 8. [lspacer]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_values: [0.5]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_value: 0.5\n",
      "INFO:pygenometracks.tracksClass:plotting 9. [test phastCons-7way]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:ylim 2.3,-0.08\n",
      "INFO:pygenometracks.tracksClass:plotting 10. [lspacer]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_values: [0.5]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_value: 0.5\n",
      "INFO:pygenometracks.tracksClass:plotting 11. [bedspacer repeats]\n",
      "WARNING:pygenometracks.tracks.GenomeTrack:*Warning*\n",
      "Neither chrX nor X existss as a chromosome name inside the bed file. This will generate an empty track!!\n",
      "\n",
      "INFO:pygenometracks.tracksClass:plotting 12. [test repeats-Alu]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:ylim 2.3,-0.08\n",
      "INFO:pygenometracks.tracksClass:plotting 13. [lspacer]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_values: [0.5]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_value: 0.5\n",
      "INFO:pygenometracks.tracksClass:plotting 14. [test repeats-Simple]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:ylim 2.3,-0.08\n",
      "INFO:pygenometracks.tracksClass:plotting 15. [lspacer]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_values: [0.5]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_value: 0.5\n",
      "# Wrote /home/hugo/software/ezTracks/test_output/test_trans_ff/my_tracks.png\n",
      "#######################################################################\n",
      "### If you want to customize the image modify /home/hugo/software/ezTracks/test_output/test_trans_ff/config.ini and then rerun\n",
      "pyGenomeTracks --tracks /home/hugo/software/ezTracks/test_output/test_trans_ff/config.ini --region chrX:73820576-73852844 --fontSize 12 --trackLabelFraction 0.2 --trackLabelHAlign left --width 25 --dpi 100 -o /home/hugo/software/ezTracks/test_output/test_trans_ff/my_tracks.png\n",
      "#######################################################################\n",
      "100%|██████████| 7/7 [00:00<00:00, 25709.39it/s]\n",
      "INFO:pygenometracks.tracksClass:initialize 1. [x-axis]\n",
      "INFO:pygenometracks.tracksClass:initialize 2. [test gtf]\n",
      "100%|██████████| 1/1 [00:00<00:00, 1563.29it/s]\n",
      "INFO:pygenometracks.tracksClass:initialize 3. [spacer]\n",
      "INFO:pygenometracks.tracksClass:initialize 5. [test CRS-CRS]\n",
      "100%|██████████| 27/27 [00:00<00:00, 34877.18it/s]\n",
      "INFO:pygenometracks.tracksClass:initialize 6. [lspacer]\n",
      "INFO:pygenometracks.tracksClass:initialize 7. [bedspacer phastCons]\n",
      "0it [00:00, ?it/s]\n",
      "WARNING:pygenometracks.tracks.GenomeTrack:No valid intervals were found in file /home/hugo/software/ezTracks/test_output/test_trans_ni/prep/empty.bed\n",
      "INFO:pygenometracks.tracksClass:initialize 8. [test phastCons-20way]\n",
      "100%|██████████| 51/51 [00:00<00:00, 36855.53it/s]\n",
      "INFO:pygenometracks.tracksClass:initialize 9. [lspacer]\n",
      "INFO:pygenometracks.tracksClass:initialize 10. [test phastCons-7way]\n",
      "100%|██████████| 19/19 [00:00<00:00, 34558.45it/s]\n",
      "INFO:pygenometracks.tracksClass:initialize 11. [lspacer]\n",
      "INFO:pygenometracks.tracksClass:initialize 12. [bedspacer repeats]\n",
      "0it [00:00, ?it/s]\n",
      "WARNING:pygenometracks.tracks.GenomeTrack:No valid intervals were found in file /home/hugo/software/ezTracks/test_output/test_trans_ni/prep/empty.bed\n",
      "INFO:pygenometracks.tracksClass:initialize 13. [test repeats-Alu]\n",
      "100%|██████████| 1/1 [00:00<00:00, 6384.02it/s]\n",
      "INFO:pygenometracks.tracksClass:initialize 14. [lspacer]\n",
      "INFO:pygenometracks.tracksClass:initialize 15. [test repeats-Simple]\n",
      "100%|██████████| 5/5 [00:00<00:00, 21183.35it/s]\n",
      "INFO:pygenometracks.tracksClass:initialize 16. [lspacer]\n",
      "INFO:pygenometracks.tracksClass:time initializing track(s):\n",
      "INFO:pygenometracks.tracksClass:0.02739715576171875\n",
      "DEBUG:pygenometracks.tracksClass:Figure size in cm is 25.0 x 9.97340425531915. Dpi is set to 100\n",
      "\n",
      "INFO:pygenometracks.tracksClass:plotting 1. [x-axis]\n",
      "INFO:pygenometracks.tracksClass:plotting 2. [test gtf]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:ylim 1.08,-0.08\n",
      "INFO:pygenometracks.tracksClass:plotting 3. [spacer]\n",
      "INFO:pygenometracks.tracksClass:plotting 5. [test CRS-CRS]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:ylim 2.3,-0.08\n",
      "INFO:pygenometracks.tracksClass:plotting 6. [lspacer]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_values: [0.5]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_value: 0.5\n",
      "INFO:pygenometracks.tracksClass:plotting 7. [bedspacer phastCons]\n",
      "WARNING:pygenometracks.tracks.GenomeTrack:*Warning*\n",
      "Neither chrX nor X existss as a chromosome name inside the bed file. This will generate an empty track!!\n",
      "\n",
      "INFO:pygenometracks.tracksClass:plotting 8. [test phastCons-20way]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:ylim 2.3,-0.08\n",
      "INFO:pygenometracks.tracksClass:plotting 9. [lspacer]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_values: [0.5]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_value: 0.5\n",
      "INFO:pygenometracks.tracksClass:plotting 10. [test phastCons-7way]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:ylim 2.3,-0.08\n",
      "INFO:pygenometracks.tracksClass:plotting 11. [lspacer]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_values: [0.5]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_value: 0.5\n",
      "INFO:pygenometracks.tracksClass:plotting 12. [bedspacer repeats]\n",
      "WARNING:pygenometracks.tracks.GenomeTrack:*Warning*\n",
      "Neither chrX nor X existss as a chromosome name inside the bed file. This will generate an empty track!!\n",
      "\n",
      "INFO:pygenometracks.tracksClass:plotting 13. [test repeats-Alu]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:ylim 2.3,-0.08\n",
      "INFO:pygenometracks.tracksClass:plotting 14. [lspacer]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_values: [0.5]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_value: 0.5\n",
      "INFO:pygenometracks.tracksClass:plotting 15. [test repeats-Simple]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:ylim 2.3,-0.08\n",
      "INFO:pygenometracks.tracksClass:plotting 16. [lspacer]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_values: [0.5]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_value: 0.5\n",
      "# Wrote /home/hugo/software/ezTracks/test_output/test_trans_ni/my_tracks.png\n",
      "#######################################################################\n",
      "### If you want to customize the image modify /home/hugo/software/ezTracks/test_output/test_trans_ni/config.ini and then rerun\n",
      "pyGenomeTracks --tracks /home/hugo/software/ezTracks/test_output/test_trans_ni/config.ini --region chrX:0-19245 --fontSize 12 --trackLabelFraction 0.2 --trackLabelHAlign left --width 25 --dpi 100 -o /home/hugo/software/ezTracks/test_output/test_trans_ni/my_tracks.png\n",
      "#######################################################################\n",
      "100%|██████████| 7/7 [00:00<00:00, 17957.26it/s]\n",
      "INFO:pygenometracks.tracksClass:initialize 1. [x-axis]\n",
      "INFO:pygenometracks.tracksClass:initialize 2. [test gtf]\n",
      "100%|██████████| 1/1 [00:00<00:00, 1172.25it/s]\n",
      "INFO:pygenometracks.tracksClass:initialize 3. [spacer]\n",
      "INFO:pygenometracks.tracksClass:initialize 5. [test CRS-CRS]\n",
      "100%|██████████| 27/27 [00:00<00:00, 34100.03it/s]\n",
      "INFO:pygenometracks.tracksClass:initialize 6. [lspacer]\n",
      "INFO:pygenometracks.tracksClass:initialize 7. [bedspacer phastCons]\n",
      "0it [00:00, ?it/s]\n",
      "WARNING:pygenometracks.tracks.GenomeTrack:No valid intervals were found in file /home/hugo/software/ezTracks/test_output/test_trans_ni_ff/prep/empty.bed\n",
      "INFO:pygenometracks.tracksClass:initialize 8. [test phastCons-20way]\n",
      "100%|██████████| 51/51 [00:00<00:00, 34884.13it/s]\n",
      "INFO:pygenometracks.tracksClass:initialize 9. [lspacer]\n",
      "INFO:pygenometracks.tracksClass:initialize 10. [test phastCons-7way]\n",
      "100%|██████████| 19/19 [00:00<00:00, 31239.43it/s]\n",
      "INFO:pygenometracks.tracksClass:initialize 11. [lspacer]\n",
      "INFO:pygenometracks.tracksClass:initialize 12. [bedspacer repeats]\n",
      "0it [00:00, ?it/s]\n",
      "WARNING:pygenometracks.tracks.GenomeTrack:No valid intervals were found in file /home/hugo/software/ezTracks/test_output/test_trans_ni_ff/prep/empty.bed\n",
      "INFO:pygenometracks.tracksClass:initialize 13. [test repeats-Alu]\n",
      "100%|██████████| 1/1 [00:00<00:00, 4739.33it/s]\n",
      "INFO:pygenometracks.tracksClass:initialize 14. [lspacer]\n",
      "INFO:pygenometracks.tracksClass:initialize 15. [test repeats-Simple]\n",
      "100%|██████████| 5/5 [00:00<00:00, 18961.59it/s]\n",
      "INFO:pygenometracks.tracksClass:initialize 16. [lspacer]\n",
      "INFO:pygenometracks.tracksClass:time initializing track(s):\n",
      "INFO:pygenometracks.tracksClass:0.02861642837524414\n",
      "DEBUG:pygenometracks.tracksClass:Figure size in cm is 25.0 x 9.97340425531915. Dpi is set to 100\n",
      "\n",
      "INFO:pygenometracks.tracksClass:plotting 1. [x-axis]\n",
      "INFO:pygenometracks.tracksClass:plotting 2. [test gtf]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:ylim 1.08,-0.08\n",
      "INFO:pygenometracks.tracksClass:plotting 3. [spacer]\n",
      "INFO:pygenometracks.tracksClass:plotting 5. [test CRS-CRS]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:ylim 2.3,-0.08\n",
      "INFO:pygenometracks.tracksClass:plotting 6. [lspacer]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_values: [0.5]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_value: 0.5\n",
      "INFO:pygenometracks.tracksClass:plotting 7. [bedspacer phastCons]\n",
      "WARNING:pygenometracks.tracks.GenomeTrack:*Warning*\n",
      "Neither chrX nor X existss as a chromosome name inside the bed file. This will generate an empty track!!\n",
      "\n",
      "INFO:pygenometracks.tracksClass:plotting 8. [test phastCons-20way]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:ylim 2.3,-0.08\n",
      "INFO:pygenometracks.tracksClass:plotting 9. [lspacer]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_values: [0.5]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_value: 0.5\n",
      "INFO:pygenometracks.tracksClass:plotting 10. [test phastCons-7way]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:ylim 2.3,-0.08\n",
      "INFO:pygenometracks.tracksClass:plotting 11. [lspacer]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_values: [0.5]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_value: 0.5\n",
      "INFO:pygenometracks.tracksClass:plotting 12. [bedspacer repeats]\n",
      "WARNING:pygenometracks.tracks.GenomeTrack:*Warning*\n",
      "Neither chrX nor X existss as a chromosome name inside the bed file. This will generate an empty track!!\n",
      "\n",
      "INFO:pygenometracks.tracksClass:plotting 13. [test repeats-Alu]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:ylim 2.3,-0.08\n",
      "INFO:pygenometracks.tracksClass:plotting 14. [lspacer]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_values: [0.5]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_value: 0.5\n",
      "INFO:pygenometracks.tracksClass:plotting 15. [test repeats-Simple]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:ylim 2.3,-0.08\n",
      "INFO:pygenometracks.tracksClass:plotting 16. [lspacer]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_values: [0.5]\n",
      "DEBUG:pygenometracks.tracks.GenomeTrack:y_value: 0.5\n",
      "# Wrote /home/hugo/software/ezTracks/test_output/test_trans_ni_ff/my_tracks.png\n",
      "#######################################################################\n",
      "### If you want to customize the image modify /home/hugo/software/ezTracks/test_output/test_trans_ni_ff/config.ini and then rerun\n",
      "pyGenomeTracks --tracks /home/hugo/software/ezTracks/test_output/test_trans_ni_ff/config.ini --region chrX:0-19245 --fontSize 12 --trackLabelFraction 0.2 --trackLabelHAlign left --width 25 --dpi 100 -o /home/hugo/software/ezTracks/test_output/test_trans_ni_ff/my_tracks.png\n",
      "#######################################################################\n"
     ]
    }
   ],
   "source": [
    "config_files = ['test/config.region.ini',\n",
    "                'test/config.trans.ini','test/config.trans_ff.ini',\n",
    "                'test/config.trans_ni.ini','test/config.trans_ni_ff.ini']\n",
    "\n",
    "#config_files = ['test/config.trans_ni_ff.ini']\n",
    "\n",
    "for config_file in config_files:\n",
    "    #args = ['eztracks.py','check',config_file]\n",
    "    #args = ['eztracks.py','prepare',config_file]\n",
    "    args = ['eztracks.py','draw',config_file]\n",
    "    sys.argv = args\n",
    "    main()    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "test/config.region.ini Mode.REGION\n",
      "test/config.trans.ini Mode.TRANS\n",
      "test/config.trans_ni.ini Mode.TRANS_NOINTRONS\n",
      "test/config.trans_ni_rn.ini Mode.TRANS_NOINTRONS_RN\n"
     ]
    }
   ],
   "source": [
    "class Mode(enum.Flag):\n",
    "    REGION = enum.auto()\n",
    "    TRANS = enum.auto()\n",
    "    TRANS_NOINTRONS = enum.auto()\n",
    "    TRANS_NOINTRONS_RN = enum.auto()\n",
    "    ERROR = enum.auto()\n",
    "    \n",
    "def get_mode(config_obj):\n",
    "    config = config_obj\n",
    "    if 'region' in config['default']:\n",
    "        return Mode.REGION\n",
    "    transcript = 'transcript' in config['default']\n",
    "    no_introns = config.getboolean('default','no_introns',fallback=False)\n",
    "    reverse_negative = config.getboolean('default','reverse_negative',fallback=False)\n",
    "    if transcript and no_introns and reverse_negative:\n",
    "        return Mode.TRANS_NOINTRONS_RN\n",
    "    if transcript and no_introns:\n",
    "        return Mode.TRANS_NOINTRONS\n",
    "    if transcript:\n",
    "        return Mode.TRANS\n",
    "    return Mode.ERROR\n",
    "\n",
    "config_files = ['test/config.region.ini','test/config.trans.ini',\n",
    "                'test/config.trans_ni.ini','test/config.trans_ni_rn.ini']\n",
    "for config_ini in config_files:\n",
    "    config, mode = load_config(config_ini)\n",
    "    print(config_ini,get_mode(config))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
